File: leetCode/python/p11_ContainerWithMostWater/lc11.py
import time

class Solution:
    """
    Given n non-negative integers a1, a2, ..., an , where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two 
    endpoints of the line i is at (i, ai) and (i, 0). Find two lines, which, together with the x-axis forms a container, such that the container 
    contains the most water.
    """
    def maxArea(self, height) -> int:
        max_area = 0
        left = 0
        right = len(height) - 1
        
        while left < right:
            area = (right - left) * min(height[left], height[right])
            
            if area > max_area:
                max_area = area
            
            if height[left] > height[right]:
                right -= 1
            else:
                left += 1

        return max_area 

    def test_function(self, height, sol: int, test_num: int): 
        start_time = time.time()
        temp = self.maxArea(height)
        end_time = time.time()
        
        if temp == sol:
            print('\nTEST' + str(test_num) + ': PASSED IN ' + '{:.2f}us'.format((end_time - start_time) * 1000000))
        else:
            print('\nTEST FAILED')        
    
s = Solution()

# Test 1

height = [1,8,6,2,5,4,8,3,7]
s.test_function(height, 49, 1)

# Test 2

height = [1,1]
s.test_function(height, 1, 2)

# Test 3

height = [4,3,2,1,4]
s.test_function(height, 16, 3)

# Test 4

height = [1,2,1]
s.test_function(height, 2, 4)

File: leetCode/c/p11_ContainerWithMostWater/lc11.c
/* Given n non-negative integers a1, a2, ..., an , where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two end    
points of the line i is at (i, ai) and (i, 0). Find two lines, which, together with the x-axis forms a container, such that the container contains     
the most water. */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))

int maxArea(int* height, int heightSize) {
    
    int max = 0, area = 0, left = 0, right = heightSize - 1;

    while(left < right) {
        area = (right - left) * MIN(height[left], height[right]);

        if(area > max) 
            max = area;

        if(height[left] > height[right])
            right--;
        else 
            left++; 
    }

    return max;
}

/* Test */ 


void testFunction(int* height, int heightSize, int sol, int testNumber) {
    
    int tmp; 
    double timeSpent;  
 
    clock_t tic = clock();
 
    tmp = maxArea(height, heightSize);
    
    clock_t toc = clock();
    
    timeSpent = (double) (toc - tic) * 1000000 / CLOCKS_PER_SEC;

    printf("\nTEST%d: ", testNumber);

    if(tmp == sol) 
        printf("PASSED IN %.2fus\n", timeSpent);
    else
        printf("FAILED\n");
}
    
int main(void) {
    
    int height1[] = {1,8,6,2,5,4,8,3,7};
    int heightSize = sizeof(height1)/sizeof(height1[0]);
    
    testFunction(height1, heightSize, 49, 1); 
    
    int height2[] = {1,1};
    heightSize = sizeof(height2)/sizeof(height2[0]);
    
    testFunction(height2, heightSize, 1, 2); 

    int height3[] = {4,3,2,1,4};
    heightSize = sizeof(height3)/sizeof(height3[0]);
    
    testFunction(height3, heightSize, 16, 3); 

    int height4[] = {1,2,1};
    heightSize = sizeof(height4)/sizeof(height4[0]);
    
    testFunction(height4, heightSize, 2, 4); 
 
    return EXIT_SUCCESS;
}
